### PR TITLE
 v2 func output data

### PR DESCRIPTION
**Pull request by anaruse**
This PR is related to a new graph representation structure (and will be related to re-compute and OOC features for memory usage reduction in future). 
Looking at [relu.py](https://github.com/pfnet/chainer/blob/_v2/chainer/functions/activation/relu.py), it has been modified using new methods retain_inputs() and retain_outputs() introduced from v2, and a reference to the output is not explicitly set in the function unlike `self.y = y` in v1. I like this approach in terms of realizing consistent data management way.
Along this approach, I've modified some activation and pooling functions in that `self.y = y` is used to retain the output until the backward, using newly introduced retain_outputs() and instance variable output_data.
I also changed a place where `func.output_data` gets set from __call__() in function.py to backward() in variable.py. The change does not give any clear advantage for now, but it will give a opportunity to further reduce memory usage when re-compute or OOC feature is implemented.

### CODE REVIEW

**Comment by anaruse** [comment]
I've added support of aggressive buffer release (#2579) to the following activation functions: leaky_relu, log_softmax, sigmoid, softmax and tanh.

File: chainer/variable.py [hunk 1]
```diff
@@ -655,8 +655,12 @@ def add_cand(cand):
             cuda.get_device(*(in_data + out_grad)).use()
             for hook in six.itervalues(hooks):
                 hook.backward_preprocess(func, in_data, out_grad)
+            func.output_data = tuple(
+                [None if y is None else y.data for y in outputs])
             gxs = func.backward(in_data, out_grad)
             assert len(gxs) == len(in_data)
+            if not hasattr(func, '_retain_after_backward'):
```
Comment (unnonouno) [inline chainer/variable.py]: `if not getattr(func, '_retain_after_backward', False)`?

File: tests/chainer_tests/test_function.py [hunk 2]
```diff
@@ -405,7 +405,7 @@ class FunctionWithRetaining(chainer.Function):
 
     def forward(self, inputs):
         self.retain_inputs([1])
-        self.retain_outputs([1])
+        self.retain_outputs([1], retain_after_backward=True)
```
Comment (unnonouno) [inline tests/chainer_tests/test_function.py]: Please make tests for botch `retain_after_backward=True` and `False`, and check if `output_data` is removed or not.

File: chainer/functions/activation/softmax.py [hunk 3]
```diff
@@ -36,37 +36,43 @@ def forward(self, x):
...
+        self._x_shape = x[0].shape
+        self._x_dtype = x[0].dtype
+        self.retain_inputs(())
+        self.retain_outputs((0,))
+        return y,
 
     def backward(self, x, gy):
-        xp = cuda.get_array_module(*x)
+        y = self.output_data[0]
+        xp = self._x_xp
```
Comment (unnonouno) [inline chainer/functions/activation/softmax.py]: Is `cuda.get_array_module(*gy)` enough?

File: chainer/functions/activation/softmax.py [hunk 4]
```diff
@@ -36,37 +36,43 @@ def forward(self, x):
...
         if (xp is not numpy and chainer.should_use_cudnn('>=auto') and
-                (_cudnn_version >= 3000 or x[0].dtype != numpy.float16)):
-            oz_dtype = 'd' if x[0].dtype == 'd' else 'f'
+                (_cudnn_version >= 3000 or self._x_dtype != numpy.float16)):
+            oz_dtype = 'd' if y[0].dtype == 'd' else 'f'
```
Comment (unnonouno) [inline chainer/functions/activation/softmax.py]: `y.dtype`

File: chainer/function.py [hunk 5]
```diff
@@ -491,7 +490,7 @@ def retain_inputs(self, indexes):
         """
         self._input_indexes_to_retain = indexes
 
-    def retain_outputs(self, indexes):
+    def retain_outputs(self, indexes, retain_after_backward=False):
```
Comment (unnonouno) [inline chainer/function.py]: Fix docstring

File: chainer/variable.py (hunk 1 above)
Comment (anaruse) [inline chainer/variable.py]: will change to use getattr() instead of hasattr(). BTW, is it better not to use hasattr()?

File: chainer/function.py (hunk 5 above)
Comment (anaruse) [inline chainer/function.py]: done

Comment (anaruse) [inline chainer/function.py]: done

File: tests/chainer_tests/test_function.py (hunk 2 above)
Comment (anaruse) [inline tests/chainer_tests/test_function.py]: Added tests both for retain_after_backward=True and False.

File: chainer/functions/activation/softmax.py (hunk 3 above)
Comment (anaruse) [inline chainer/functions/activation/softmax.py]: Yes. Fixed code based on your comment.

File: chainer/functions/activation/softmax.py (hunk 4 above)
Comment (anaruse) [inline chainer/functions/activation/softmax.py]: done

File: chainer/function.py [hunk 6]
```diff
@@ -514,8 +513,14 @@ def retain_outputs(self, indexes):
             indexes (iterable of int): Indexes of input variables that the
                 function does not require for backprop.
 
+            retain_after_backward (bool): If ``True``, a reference to the
+                outputs wil remain after the backprop of the function is over.
```
Comment (anaruse) [inline chainer/function.py]: fixed

**Comment by beam2d** [comment]
LGTM. Sorry for being late, but in order to include this PR to v2, we have to finish the merge today. @anaruse Do you have time to resolve the merge conflicts today? If not, we would like to do the merge by ourselves.

**Comment by beam2d** [comment]
Based on the discussion with @okuta, I found that removing the output data during backward prevents users from doing backward multiple times for a single computational graph. It would be problematic, and it seems that it is better to let users instead of each function implementation decide whether or not to retain the data on backward. I think it's better to have more discussions on this point and make this change after v2.

**Comment by anaruse** [comment]
I understood. I will update the branch so that the references to the output data (func.output_data) are not removed at default even after the backprop.
Anyway, I've resolved the conflicts to v2 master.

**Comment by beam2d** [comment]
Thank you for the quick response! After making that change, I expect it can be included in the next version.

**Comment by anaruse** [comment]
Probably, this PR is safe as it is even when a computational graph is back-proped multiple times.

What this PR does in backward() of variable.py is that 1) to make a reference to output data of the function in _**Function**_ as **func.output** before the **func.backward()** is called and 2) to remove the reference **func.output** after the **func.backward()** is done without changing any instances in _**Function**_ and _**VariableNode**_ that are related to computational graph structure. Since the computational graphs is kept unchanged, the processing above should work correctly even in 2nd or later backprop to that computation graph.

Is my understanding wrong?

**Comment by beam2d** [comment]
Ah, I now understood what this PR actually does. OK, then it seems that this code works for multiple backprop calls.

**Comment by beam2d** [comment]
I found the case that the `retain_after_backward=False` may cause a bug (I have not noticed it...):

```
import chainer
import numpy as np

class Func(chainer.Function):
    def forward(self, xs):
        self.retain_outputs((0, 1))
        return xs

    def backward(self, xs, gys):
        print(self.output_data)
        return gys  # it might be replaced by an expression that uses output_data

def func(x, y):
    return Func()(x, y)

x = chainer.Variable(np.array([1.]))
y = chainer.Variable(np.array([2.]))
z, w = func(x, y)
z.backward()  # print: (array([ 1.]), array([ 2.]))
              # At this point, both of retained outputs are accessible
del w
z.backward()  # print: (array([ 1.]), None)
              # Here w is inaccessible even from backward()
```

That is, when an output variable has been released before backward is called, the output array cannot be recovered even if we let the function retain the output during forward prop. I think it should be fixed.

**Comment by anaruse** [comment]
Ah, you are right. In that case, the data retained by `retain_outputs` gets invisible during backward. I suppose this happens when the `variable node` of the data is a leaf node in computational graph and the `variable` of the data is freed. Now, I'm thinking how to fix/work-around this...

I'm not sure
